vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java

```diff
@@ -390,4 +448,151 @@ Collection<Component> getDirectChildren() {
+                return false;
+            }
+            // Assert section contents
+            if (childMatch.get() instanceof DashboardSection sectionMatch
+                    && !areSectionItemsInSync(sectionMatch,
+                            rootLevelItemFromClient, allNodeIds)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean areSectionItemsInSync(DashboardSection section,
+            JsonObject sectionFromClient, Set<Integer> allNodeIds) {
+        // Assert section has items array
+        if (!sectionFromClient.hasKey("items")) {
+            return false;
+        }
+        JsonArray sectionItemsFromClient = sectionFromClient.getArray("items");
+        // Assert section child item count
+        if (sectionItemsFromClient.length() != section.getWidgets().size()) {
+            return false;
+        }
+        for (int sectionItemIdx = 0; sectionItemIdx < sectionItemsFromClient
+                .length(); sectionItemIdx++) {
+            int sectionItemNodeId = (int) sectionItemsFromClient
+                    .getObject(sectionItemIdx).getNumber("nodeid");
+            // Assert unique node ID
+            if (allNodeIds.contains(sectionItemNodeId)) {
+                return false;
+            }
+            allNodeIds.add(sectionItemNodeId);
+            // Assert section child item node ID
+            if (section.getWidgets().stream().noneMatch(child -> child
+                    .getElement().getNode().getId() == sectionItemNodeId)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void customizeItemReorderEndEvent() {
+        getElement().executeJs(
+                """
+                        this.addEventListener('dashboard-item-reorder-end', (e) => {
+                          const itemsCopy = [];
+                          for (let rootLevelIdx = 0; rootLevelIdx < this.items; rootLevelIdx++) {
+                            const item = this.items[rootLevelIdx];
+                            const itemCopy = { nodeid: item.nodeid };
+                            if (item.items) {
+                              const sectionItemsCopy = [];
+                              const sectionItems = item.items;
+                              for (let sectionItemIdx = 0; sectionItemIdx < sectionItems; sectionItemIdx++) {
+                                const sectionItemCopy = { nodeid: sectionItems[sectionItemIdx].nodeid };
+                                sectionItemsCopy.push(sectionItemCopy);
+                              }
+                              itemCopy.items = sectionItemsCopy;
+                            }
+                            itemsCopy.push(itemCopy);
+                          }
+                          const flowReorderEvent = new CustomEvent('dashboard-item-reorder-end-flow', {
+                            detail: { items: itemsCopy },
+                          });
```

Comment:
Could shorten this to:
```js
function mapItems(items) {
  return items.map(({nodeid, items}) => ({
    nodeid,
    ...(items && { items: mapItems(items) })
  }));
}

const flowReorderEvent = new CustomEvent('dashboard-item-reorder-end-flow', {
  detail: { items: mapItems(this.items) }
});
```

Reply:
Updated the code to use this shortened version.